### PR TITLE
tests: enable multiple pageservers in `neon_local` and `neon_fixture`

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Currently our testing environment only supports running a single pageserver at a time.  This is insufficient for testing failover and migrations.
- Dependency of writing tests for #5207 

## Summary of changes

- `neon_local` and `neon_fixture` now handle multiple pageservers
- This is a breaking change to the `.neon/config` format: any local environments will need recreating
- Existing tests continue to work unchanged:
  - The default number of pageservers is 1
  - `NeonEnv.pageserver` is now a helper property that retrieves the first pageserver if there is only one, else throws.
- Pageserver data directories are now at `.neon/pageserver_{n}` where n is 1,2,3...
- Compatibility tests get some special casing to migrate neon_local configs: these are not meant to be backward/forward compatible, but they were treated that way by the test.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: test_runner/regress/test_timeline_delete.py [hunk 1]
```diff
@@ -86,7 +90,11 @@ def test_timeline_delete(neon_simple_env: NeonEnv):
     assert exc.value.status_code == 412
 
     timeline_path = (
-        env.repo_dir / "tenants" / str(env.initial_tenant) / "timelines" / str(leaf_timeline_id)
+        env.pageserver.workdir
+        / "tenants"
+        / str(env.initial_tenant)
+        / "timelines"
+        / str(leaf_timeline_id)
```
[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_timeline_delete.py:97]: Perhaps we should upgrade all of these to use a variant of that?

[the remark this answers is not in the document]
Comment (jcsp) [reply test_runner/regress/test_timeline_delete.py:97]: Yep, `timeline_dir` was updated.

For replacing these with a helper, I think we should do that in the future at the same time as refactoring the pageserver path helpers into NeonPageserver, out of the NeonEnv monolith

[the remark this answers is not in the document]
Comment (koivunej) [reply test_runner/regress/test_timeline_delete.py:97]: Created: https://github.com/neondatabase/neon/issues/5258

**Review by koivunej — APPROVED** [review]
I haven't found anything else to nit, already created an issue for my only thread.

File: test_runner/fixtures/neon_fixtures.py [hunk 2]
```diff
@@ -1564,8 +1635,12 @@ def http_client(self, auth_token: Optional[str] = None) -> PageserverHttpClient:
             is_testing_enabled_or_skip=self.is_testing_enabled_or_skip,
         )
 
+    @property
+    def workdir(self) -> Path:
+        return Path(os.path.join(self.env.repo_dir, f"pageserver_{self.id}"))
```
Comment (bayandin) [inline test_runner/fixtures/neon_fixtures.py:1640]: I think `repo_dir` is a `Path`

```suggestion
        return self.env.repo_dir / f"pageserver_{self.id}"
```

**Review by bayandin — APPROVED** [review]
Great work!
Just a small comment

File: test_runner/fixtures/neon_fixtures.py (hunk 2 above)
Comment (jcsp) [reply test_runner/fixtures/neon_fixtures.py:1640]: You're right, I'll put a little commit in my next PR that touches this file, to pick it up (the next PR is the one that actually uses multiple pageservers for real)

**Pull request merged**